### PR TITLE
Fix element order in Python wrapping of circuit accessors.

### PR DESCRIPTION
**Pull request by hernando**
The new implementation preserve the iteration order of the python
object used as input GID set.

The functions that were previously returning Python sets for GID sets
now return sorted numpy arrays. It should still be easy to perform set
operations using numpy:
https://docs.scipy.org/doc/numpy/reference/routines.set.html

### CODE REVIEW

File: brain/python/arrayHelpers.cpp [hunk 1]
```diff
@@ -195,6 +200,32 @@ bp::object toNumpy( const brain::Matrix4f& matrix )
...
+    for( size_t i = 0; i != size; ++i )
+    {
+        const T gid = *static_cast< T* >(PyArray_GETPTR1( array, i ));
+        if( gid < 0 || ssize_t( gid ) >
+                       ssize_t( std::numeric_limits< uint32_t >::max( )))
+            PyErr_SetString( PyExc_ValueError, "Invalid input GID" );
+        if( last >= gid )
+            sorted = false;
+        else
+            sorted = gid;
```
Comment (tribal-tec) [inline brain/python/arrayHelpers.cpp]: couldn't this if/else collapse to ``` sorted = last < gid;``` ?

File: brain/python/arrayHelpers.cpp [hunk 2]
```diff
@@ -195,6 +200,32 @@ bp::object toNumpy( const brain::Matrix4f& matrix )
 
 namespace
 {
+
+template< typename T >
+bool _copyGIDs( PyArrayObject* array, uint32_ts& result )
+{
+    size_t size = PyArray_DIMS(array)[0];
```
Comment (tribal-tec) [inline brain/python/arrayHelpers.cpp]: const

File: brain/python/arrayHelpers.h [hunk 3]
```diff
@@ -30,24 +30,37 @@ namespace brain
 
 void importArray();
 
+/** @return True is the python object is a numpy.ndarray, false otherwise. */
```
Comment (tribal-tec) [inline brain/python/arrayHelpers.h]: replace 'is' with 'if'

File: brain/python/helpers.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        // from an iterable
+        boost::python::stl_input_iterator< unsigned int > i( iterable ), end;
+        uint32_t last = 0;
+        for( ; i != end; ++i )
+        {
+            const uint32_t gid = *i;
+            if( last >= gid )
+                sorted = false;
+            else
+                last = gid;
```
Comment (tribal-tec) [inline brain/python/helpers.cpp]: same again as before (C&P?)

File: brain/python/helpers.h [hunk 5]
```diff
@@ -57,6 +59,15 @@ inline boost::python::list toPythonList( const std::vector< T >& vector )
     return result;
 }
 
+inline uint32_ts toVector( const GIDSet& gids )
```
Comment (tribal-tec) [inline brain/python/helpers.h]: why not ``` uint32_ts( gids.begin(), gids.end( )); ``` ?

File: brain/python/helpers.h [hunk 6]
```diff
@@ -70,26 +81,22 @@ inline boost::python::object toPythonSet( const GIDSet& ids )
...
-    }
-    catch(...)
-    {
-        PyErr_SetString( PyExc_ValueError,
-                         "Cannot convert argument to GID set" );
-        boost::python::throw_error_already_set();
-    }
-    return result;
-}
+/** Copy the contents of a Python iterable into or numpy array into a GIDSet. */
```
Comment (tribal-tec) [inline brain/python/helpers.h]: ...into or ?

File: doc/Changelog.md [hunk 7]
```diff
@@ -3,6 +3,11 @@ Changelog {#Changelog}
 
 # git master
 
+* [115](https://github.com/BlueBrain/Brion/pull/113):
+  Changes and fixes in brian Python module:
```
Comment (tribal-tec) [inline doc/Changelog.md]: who is brian? :)

File: doc/Changelog.md [hunk 8]
```diff
@@ -3,6 +3,11 @@ Changelog {#Changelog}
 
 # git master
 
+* [115](https://github.com/BlueBrain/Brion/pull/113):
```
Comment (tribal-tec) [inline doc/Changelog.md]: 117 two times please

File: brion/circuit.cpp [hunk 9]
```diff
@@ -103,7 +103,7 @@ class Circuit::Impl
         indices.reserve( gids.size( ));
         BOOST_FOREACH( const uint32_t gid, gids )
```
Comment (tribal-tec) [inline brion/circuit.cpp]: C++11 for-loop

File: doc/Changelog.md (hunk 7 above)
Comment (tribal-tec) [inline doc/Changelog.md]: Also mention the API change in Circuit wrt m-/e- types function names

File: brain/python/arrayHelpers.cpp (hunk 1 above)
Comment (hernando) [inline brain/python/arrayHelpers.cpp]: There's a mistake here, sorted = gid should be last = gid

File: brain/python/helpers.h (hunk 5 above)
Comment (hernando) [inline brain/python/helpers.h]: See lines 85-101 in brain/python/circuit.cpp, without the helper function you need a temporary variable.

File: brain/python/helpers.h (hunk 6 above)
Comment (hernando) [inline brain/python/helpers.h]: ~into~

File: doc/Changelog.md (hunk 7 above)
Comment (hernando) [inline doc/Changelog.md]: That reminds me that I have to build everything to see which code the change breaks.

File: brain/python/helpers.h (hunk 5 above)
Comment (tribal-tec) [inline brain/python/helpers.h]: but you can still implement toVector() with the above mentioned constructor, no?

File: brain/python/helpers.h [hunk 10]
```diff
@@ -70,26 +81,22 @@ inline boost::python::object toPythonSet( const GIDSet& ids )
...
+GIDSet gidsFromPython( const boost::python::object& object );
+
+/** Copy the contents of a Python iterable or numpy array into a GIDSet and
+    return the correspondance map between elements in the iterable and the set.
+    @param object A Python iterable object or numpy.ndarray
+    @param result The out GID set.
+    @param result mapping If the input iterable is not sorted, this vector
```
Comment (rdumusc) [inline brain/python/helpers.h]: rm result

File: brain/python/arrayHelpers.h (hunk 3 above)
Comment (hernando) [inline brain/python/arrayHelpers.h]: Yes, but why didn't you see the latest changes? Also most comments from Daniel are outdated by they still appear.

File: brain/python/arrayHelpers.cpp [hunk 11]
```diff
@@ -195,6 +200,32 @@ bp::object toNumpy( const brain::Matrix4f& matrix )
...
+    result.clear();
+    result.reserve( size );
+
+    T last = 0;
+    for( size_t i = 0; i != size; ++i )
+    {
+        const T gid = *static_cast< T* >(PyArray_GETPTR1( array, i ));
+        if( gid < 0 || ssize_t( gid ) >
+                       ssize_t( std::numeric_limits< uint32_t >::max( )))
+            PyErr_SetString( PyExc_ValueError, "Invalid input GID" );
```
Comment (hernando) [inline brain/python/arrayHelpers.cpp]: Good catch. I was not testing this explicitly.

File: brain/python/arrayHelpers.cpp (hunk 1 above)
Comment (hernando) [inline brain/python/arrayHelpers.cpp]: already done

File: brain/python/arrayHelpers.cpp (hunk 2 above)
Comment (hernando) [inline brain/python/arrayHelpers.cpp]: done

File: brain/python/arrayHelpers.h (hunk 3 above)
Comment (rdumusc) [inline brain/python/arrayHelpers.h]: I had started before, and finished after a discussion with Nicolas ;-)

Comment (hernando) [inline brain/python/arrayHelpers.h]: done

File: brain/python/helpers.h (hunk 5 above)
Comment (hernando) [inline brain/python/helpers.h]: done

Comment (tribal-tec) [inline brain/python/helpers.h]: where? I don't see it in CR and CR2 commits.

Comment (hernando) [inline brain/python/helpers.h]: ?? Check again, it should be in CR2

**Pull request merged**